tests/run_test.sh

```diff
@@ -0,0 +1,130 @@
+#!/bin/sh
+
+usage() {
+  cat >&2 << HELP
+Usage: run_test [--bless] TESTS...
+Run tests from the test suite.
+
+Each TEST should be the name of a TEST.in file in the current directory.
+It calls 'metamath TEST.mm < TEST.in > TEST.produced' and compares
+TEST.produced with TEST.expected, printing a test failure report.
+
+If TEST.mm does not exist, then it just calls metamath without arguments.
+
+The 'metamath' command can be modified by setting the METAMATH environment
+variable.
+
+The --bless option can be used to update the TEST.expected file to match
+TEST.produced. Always review the changes after a call to run_test --bless.
+
+TEST.in files have the syntax of metamath scripts, so leading ! can be used to
+write comments. This script contains some special directives in tests:
+
+* '! run_test' means that the output will be ignored: TEST.produced will not be
+  created and TEST.expected does not have to exist.
+* '! expect_fail' means that we expect metamath to return exit code 1
+  instead of 0 on this input.
+HELP
+}
+
+if [ "$1" = "--help" ]; then usage; exit; fi
+if [ "$1" = "--bless" ]; then bless=1; shift; fi
+
+# Allow overriding the 'metamath' command using the METAMATH env variable
+cmd="${METAMATH:-metamath}"
```

Comment:
I consider this kind of parameter submission troublesome, because setting/unsetting this variable may differ both in location and time from its usage here.  Better submit it as a call parameter.

Reply:
I think it is conventional to use environment variables for command substitution, for example with the `$(CC)` variable in make to change the compiler. I will add both methods.

Reply:
Actually I'm not sure I see any difference this way. There are two ways (now three) to use the script:

1. `./run_test.sh test`
2. `METAMATH=path/to/metamath ./run_test.sh test`
3. `./run_test.sh -c path/to/metamath test`

With (1), intended for maximum convenience, it will use `metamath` if you have it on your path, else you can set the METAMATH env variable (perhaps earlier in your session, perhaps in your `~/.profile` and long forgotten) to control where to find the program. This version has the downside that it can break due to misconfiguration: `./run_test.sh test` might mysteriously fail to work because METAMATH env var got set somehow. (Note that the script will give an error message in this case, so it shouldn't be too hard to diagnose.)

With (2), the environment variable is being set to something, so there is no possibility of misconfiguration. This is useful for scripts and CI, but it is more verbose. You always get the command specified.

With (3), you also always get the command specified. So it seems to have the same benefits and drawbacks as (2), except for a minor syntax change. So I don't really see the point of having both.

Reply:
In our company we take care of a larger script project where the mentioned kind of "remote parameter injection" was routinely used.  This turned out to be a pain for maintenance, and was the cause for a lot of subtle bugs.

Reply:
This is a highly appreciated comment on https://stackoverflow.com/questions/484635/are-global-variables-bad.

"The problem with global variables is that since every function has access to these, it becomes increasingly hard to figure out which functions actually read and write these variables.

To understand how the application works, you pretty much have to take into account every function which modifies the global state. That can be done, but as the application grows it will get harder to the point of being virtually impossible (or at least a complete waste of time).

If you don't rely on global variables, you can pass state around between different functions as needed. That way you stand a much better chance of understanding what each function does, as you don't need to take the global state into account."

This, by the way, describes __exactly__ the maintenance problems of the script project in my company.

Or translated to the language of a software engineer:  You cannot assert the preconditions of the script any more, if you know nothing about the METAMATH environment variable, that may be accidentally set to something unexpected for whatever reason.  I think it is particularly difficult in a context, where you trust strict verification of mathematical proofs.  Convenience imo is NOT the highest priority, but predictability is.  

As for the shebang:  I don't think this is really a big issue.  I'd rather have this solved in a code style reading something along the lines:
"We allow using #!/bin/sh if the script passes a call with bash --posix.", accepted publicly by contributors X Y and Z.
In a review I can then simply look up whether the shebang matches this criterium, instead of unconditionally accepting a "I believe this won't hurt".

Reply:
> Or translated to the language of a software engineer: You cannot assert the preconditions of the script any more, if you know nothing about the METAMATH environment variable, that may be accidentally set to something unexpected for whatever reason. I think it is particularly difficult in a context, where you trust strict verification of mathematical proofs. Convenience imo is NOT the highest priority, but predictability is.

As I said, if you want to get reproducible results then you take the global state out of the equation by using the `METAMATH=path/to/metamath ./run_test.sh test` or `./run_test.sh -c path/to/metamath test` forms, and you can see that this is what is used in CI. However, reproducibility is not the most important thing in all situations. Command line tools like `cargo` and `git` need to have good defaults and configuration at multiple levels in order to enable *convenient* use. Yes, this is global state, but it's there for a reason.

Regarding `/bin/sh`: this is already symlinked to `bash --posix` on my machine and this is the default on Ubuntu machines if not others. All testing of this script, including the CI, has been running it in this mode, so it should be easy for a reviewer to verify this.

Reply:
Not quite sure whether this is clearly understood: I don't deny a well-defined call is possible if certain restrictions are obeyed. My point is that the interface allows misuse, and, thus, is inherently unsafe.  This is either seen as a problem, or, well, not.

What should be clarified, maybe not in this thread, is whether I should continue reviewing at all.  The general task of a reviewer is described here https://en.wikipedia.org/wiki/Code_review.  I further published review check lists for discussion, and laid open that, without any rules established, (see #40 first post) I follow the guidelines of my company.  I am increasingly under the impression to deliver results others are not happy with.  And from my personal perspective things don't go well either.  My expectation is that any not openly flawed review comment should result in a code style, review check, or similar, that in future makes clear, how to proceed either in coding, testing or reviewing.  In the long run, we so could arrive at a well-tested development cycle, in which many persons can participate.  I don't see any progress here...

Just some thoughts popping off the top of my head.  I will post something to this respect in issue #40, I think, hopefully within the next week.

Reply:
> My point is that the interface allows misuse, and, thus, is inherently unsafe.

The second half of this statement doesn't follow from the first half. There is a perfectly well defined contract for what's going on here. The environment variable is one of the inputs to that contract.

>  My expectation is that any not openly flawed and contentious review comment should result in a code style, review check, or similar, that in future makes clear, how to proceed either in coding, testing or reviewing. In the long run, we so could arrive at a well-tested development cycle, in which many persons can participate. I don't see any progress here...

I *did* make a change in response to this comment. I did not remove the environment variable because it is important to my workflow. Subsequently I have tried to argue for why it is there and we seem to have come to an impasse. I think we're not likely to make any more progress on this issue. If you don't trust the script, don't use it. You can verify that it is being used correctly in CI, and the other features are intended for well-meaning non-adversarial developers to be able to update the test suite.

Regarding the meta-point, I do believe our expectations are misaligned. I value your input, all of your reviews have been useful to me, but I get a general sense that you don't trust me or my work that negatively colors the review process. My experiences with [metamath-knife](https://github.com/david-a-wheeler/metamath-knife/) have gone more smoothly, even though it's mostly only @tirix who has the time to review.

Reply:
I think my last post was a lot motivated by my observation that I make close to no progress in the review and the documentation field.  IMO the review should not be governed by rules in effect in my company, that look quite arbitrary from a metamath standpoint.  But I am forced to.  Without any handout I can just follow my preferences.

These preferences in turn are based on my experience with code and where it badly fails in practice.  Relying on state not passed as a parameter is such a pitfall, and a contract given in form of a help text is a weak support.  I still don't understand why `run-tests -c "$METAMATH" <tests>` is not sufficient.

But again, my personal preferences should not be a critical factor in review.  So if a review rule states "The reviewer verifies that interfaces described by contracts hold.  S/he might annotate warnings if she detects signs of bad architecture, but nothing else.", then we wouldn't have this discussion.  Such a rule could be the spin off of this discussion, for example.